src/marks/dot.js

```diff
@@ -50,6 +75,71 @@ export class Dot extends Mark {
   }
 }
 
+function maybeDodge(dodge, x, y) {
+  if (dodge === false) dodge = null;
+  if (dodge === true) dodge = x == null ? "left" : "bottom";
+  if (dodge == null) return null;
+  switch (dodge = `${dodge}`.toLowerCase()) {
+    case "top":
+    case "bottom":
+      if (x == null) throw new Error("missing channel: x");
+      if (y != null) throw new Error("conflicting channel: y");
+      return dodge;
+    case "left":
+    case "right":
+      if (y == null) throw new Error("missing channel: y");
+      if (x != null) throw new Error("conflicting channel: x");
+      return dodge;
+    case "center":
+      if (x == null && y == null) throw new Error("missing channel: x or y");
+      if (x != null && y != null) throw new Error("conflicting channel: x and y");
+      return dodge;
+    default:
+      throw new Error(`unknown dodge method: ${dodge}`);
+  }
+}
+
+function dodge(I, X, R, Rsep, compare) {
+  const Y = new Float64Array(X.length);
+  const tree = IntervalTree();
+  for (const i of I) {
+    const intervals = [];
+    const l = X[i] - R[i];
+    const r = X[i] + R[i];
+
+    // For any previously placed circles that may overlap this circle, compute
+    // the y-positions that place this circle tangent to these other circles.
+    // https://observablehq.com/@mbostock/circle-offset-along-line
+    tree.queryInterval(l - Rsep, r + Rsep, ([,, j]) => {
+      const yj = Y[j];
+      const dx = X[i] - X[j];
+      const dr = R[i] + Rsep + R[j];
+      const dy = Math.sqrt(dr * dr - dx * dx);
+      intervals.push([yj - dy, yj + dy]);
+    });
+
+    // Find the best y-value where this circle can fit.
+    for (let y of intervals.flat().sort(compare)) {
+      if (intervals.every(([lo, hi]) => y <= lo || y >= hi)) {
+        Y[i] = y;
```

Comment:
Currently we’re setting Y[i] = 0 if there’s no conflict, but I wonder at least in the case of asymmetric swarms whether we should instead be setting it to R[i] + Rsep so that the bottom of the swarm is aligned (rather than the centers of the circles).